README.md
<br>简体中文 | [English](README_en.md)

[![CI](https://github.com/gngpp/ninja/actions/workflows/CI.yml/badge.svg)](https://github.com/gngpp/ninja/actions/workflows/CI.yml)
[![CI](https://github.com/gngpp/ninja/actions/workflows/Release.yml/badge.svg)](https://github.com/gngpp/ninja/actions/workflows/Release.yml)
 <a target="_blank" href="https://github.com/gngpp/vdns/blob/main/LICENSE">
  <img src="https://img.shields.io/badge/license-MIT-blue.svg"/>
 </a>
  <a href="https://github.com/gngpp/ninja/releases">
    <img src="https://img.shields.io/github/release/gngpp/ninja.svg?style=flat">
  </a><a href="https://github.com/gngpp/ninja/releases">
    <img src="https://img.shields.io/github/downloads/gngpp/ninja/total?style=flat">
  </a>
  [![](https://img.shields.io/docker/image-size/gngpp/ninja)](https://registry.hub.docker.com/r/gngpp/ninja)
  [![Docker Image](https://img.shields.io/docker/pulls/gngpp/ninja.svg)](https://hub.docker.com/r/gngpp/ninja/)

# ninja

逆向工程的 `ChatGPT` 代理（绕过 Cloudflare 403 Access Denied）

### 功能

- API密钥获取
- 电子邮件/密码帐户认证 (暂时不支持Google/Microsoft第三方登录)
- `Unofficial`/`Official`/`ChatGPT-to-API` Http API 代理 (供第三方客户端接入)
- 支持代理池
- 极少的内存占用
- 附带的ChatGPT WebUI

> 局限性: 无法绕过 OpenAI 的彻底 IP 禁令

### 绕过IP限制

这里`IP限制`是指`OpenAI`对`单IP`请求速率限制，你需要了解什么是`puid`，默认请求models接口返回`puid cookie`。
另外，`GPT-4`会话必须带上`puid`发送，在使用第三方客户端发送`GPT-4`会话时可能不会保存也不会获取`puid`，你需要在服务端处理:

- 使用启动参数`--puid`单独设置共享使用，此方式不支持更新
- 使用启动参数`--puid-user`，设置`Account Plus`账号来获取`puid`，并且会定时更新

### ArkoseLabs

发送`GPT4`对话需要`Arkose Token`作为参数发送，支持的解决方案暂时只有三种

1) 获取`Arkose Token`的端点，不管你用什么方式，使用 `--arkose-token-endpoint` 指定端点获取token，支持的`JSON`格式，一般按照社区的格式：`{"token":"xxxxxx"}`

2) 使用HAR，`ChatGPT` 官网发送一次 `GPT4` 会话消息，浏览器 `F12` 下载 `https://tcr9i.chat.openai.com/fc/gt2/public_key/35536E1E-65B4-4D96-9D97-6ADB7EFF8147` 接口的HAR日志记录文件，使用启动参数 `--arkose-har-file` 指定HAR文件路径使用（不指定路径则使用默认路径`~/chat.openai.com.har`，可直接上传更新HAR），支持上传更新HAR，请求路径:`/har/upload`，可选上传身份验证参数:`--arkose-har-upload-key`

3) 使用[YesCaptcha](https://yescaptcha.com/i/1Cc5i4) / [CapSolver](https://dashboard.capsolver.com/passport/register?inviteCode=y7CtB_a-3X6d)平台进行验证码解析，启动参数`--arkose-solver`选择平台（默认使用`YesCaptcha`），`--arkose-solver-key` 填写`Client Key`

- 三种方案都使用，优先级是：`HAR` > `YesCaptcha` / `CapSolver` > `Arkose Token 端点`
- `YesCaptcha` / `CapSolver`推荐搭配HAR使用，出验证码则调用解析器处理，验证后HAR使用更持久

### Command Line(dev)

### Http 服务

#### 公开接口, `*` 表示任意`URL`后缀

- backend-api, <https://host:port/backend-api/*>
- public-api, <https://host:port/public-api/*>
- platform-api, <https://host:port/v1/*>
- dashboard-api, <https://host:port/dashboard/*>
- chatgpt-to-api, <https://host:port/to/v1/chat/completions>

#### API文档

- Platfrom API [doc](https://platform.openai.com/docs/api-reference)
- Backend API [doc](doc/rest.http)

> 关于关于`ChatGPT`转`API`使用，直接拿`AceessToken`当`API Key`使用，接口路径：`/to/v1/chat/completions`

#### 基本服务

- 原汁原味ChatGPT WebUI
- 公开`非官方`/`官方API`代理
- `API`前缀与官方一致
- `ChatGPT` 转 `API`
- 可接入第三方客户端
- 可接入IP代理池，提高并发

- 参数说明
  - `--level`，环境变量 `LOG`，日志级别: 默认info
  - `--host`，环境变量 `HOST`， 服务监听地址: 默认0.0.0.0，
  - `--port`，环境变量 `PORT`， 监听端口: 默认7999
  - `--tls-cert`，环境变量 `TLS_CERT`，TLS证书公钥，支持格式: EC/PKCS8/RSA
  - `--tls-key`，环境变量 `TLS_KEY`，TLS证书私钥
  - `--proxies`，代理，支持代理池，多个代理使用`,`隔开，格式: protocol://user:pass@ip:port，如果本地IP被Ban，使用代理池时需要关闭直连IP使用，`--disable-direct`关闭直连，否则会根据负载均衡使用你被Ban的本地IP
  - `--workers`， 工作线程: 默认1
  - `--disable-webui`, 如果不想使用默认自带的WebUI，使用此参数关闭

[...](https://github.com/gngpp/ninja/blob/main/README_zh.md#%E5%91%BD%E4%BB%A4%E6%89%8B%E5%86%8C)

### 安装

- #### Ubuntu(Other Linux)

  GitHub [Releases](https://github.com/gngpp/ninja/releases/latest) 中有预编译的 deb包，二进制文件，以Ubuntu为例：

```shell
wget https://github.com/gngpp/ninja/releases/download/v0.5.6/ninja-0.5.6-x86_64-unknown-linux-musl.deb
dpkg -i ninja-0.5.6-x86_64-unknown-linux-musl.deb
ninja serve run
```

- #### OpenWrt

GitHub [Releases](https://github.com/gngpp/ninja/releases/latest) 中有预编译的 ipk 文件， 目前提供了 aarch64/x86_64 等架构的版本，下载后使用 opkg 安装，以 nanopi r4s 为例：

```shell
wget https://github.com/gngpp/ninja/releases/download/v0.5.6/ninja_0.5.6_aarch64_generic.ipk
wget https://github.com/gngpp/ninja/releases/download/v0.5.6/luci-app-ninja_1.0.9-1_all.ipk
wget https://github.com/gngpp/ninja/releases/download/v0.5.6/luci-i18n-ninja-zh-cn_1.0.9-1_all.ipk

opkg install ninja_0.5.6_aarch64_generic.ipk
opkg install luci-app-ninja_1.0.9-1_all.ipk
opkg install luci-i18n-ninja-zh-cn_1.0.9-1_all.ipk
```

- #### Docker

```shell
docker run --rm -it -p 7999:7999 --name=ninja \
  -e WORKERS=1 \
  -e LOG=info \
  gngpp/ninja:latest serve run
```

- Docker Compose

> `CloudFlare Warp`你的地区不支持（China）请把它删掉，或者你的`VPS`IP可直连`OpenAI`，那么也可以删掉

```yaml
version: '3'

services:
  ninja:
    image: ghcr.io/gngpp/ninja:latest
    container_name: ninja
    restart: unless-stopped
    environment:
      - TZ=Asia/Shanghai
      - PROXIES=socks5://warp:10000
      # - CONFIG=/serve.toml
      # - PORT=8080
      # - HOST=0.0.0.0
      # - TLS_CERT=
      # - TLS_KEY=
    # volumes:
      # - ${PWD}/ssl:/etc
      # - ${PWD}/serve.toml:/serve.toml
    command: serve run
    ports:
      - "8080:7999"
    depends_on:
      - warp

  warp:
    container_name: warp
    image: ghcr.io/gngpp/warp:latest
    restart: unless-stopped

  watchtower:
    container_name: watchtower
    image: containrrr/watchtower
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    command: --interval 3600 --cleanup
    restart: unless-stopped

```

### 命令手册

```shell
$ ninja serve --help
Start the http server

Usage: ninja serve run [OPTIONS]

Options:
  -C, --config <CONFIG>
          Configuration file path (toml format file) [env: CONFIG=]
  -H, --host <HOST>
          Server Listen host [env: HOST=] [default: 0.0.0.0]
  -L, --level <LEVEL>
          Log level (info/debug/warn/trace/error) [env: LOG=] [default: info]
  -P, --port <PORT>
          Server Listen port [env: PORT=] [default: 7999]
  -W, --workers <WORKERS>
          Server worker-pool size (Recommended number of CPU cores) [default: 1]
      --concurrent-limit <CONCURRENT_LIMIT>
          Enforces a limit on the concurrent number of requests the underlying [default: 65535]
  -x, --proxies <PROXIES>
          Server proxies pool, Example: protocol://user:pass@ip:port [env: PROXIES=]
      --disable-direct
          Disable direct connection [env: DISABLE_DIRECT=]
      --timeout <TIMEOUT>
          Client timeout (seconds) [default: 600]
      --connect-timeout <CONNECT_TIMEOUT>
          Client connect timeout (seconds) [default: 60]
      --tcp-keepalive <TCP_KEEPALIVE>
          TCP keepalive (seconds) [default: 60]
      --tls-cert <TLS_CERT>
          TLS certificate file path [env: TLS_CERT=]
      --tls-key <TLS_KEY>
          TLS private key file path (EC/PKCS8/RSA) [env: TLS_KEY=]
      --puid <PUID>
          PUID cookie value of Plus account [env: PUID=]
      --puid-user <PUID_USER>
          Obtain the PUID of the Plus account user, Example: `user:pass` or `user:pass:mfa`
      --api-prefix <API_PREFIX>
          Web UI api prefix [env: UI_API_PREFIX=]
      --arkose-endpoint <ARKOSE_ENDPOINT>
          Arkose endpoint, Example: https://client-api.arkoselabs.com
  -A, --arkose-token-endpoint <ARKOSE_TOKEN_ENDPOINT>
          Get arkose token endpoint
  -a, --arkose-har-file <ARKOSE_HAR_FILE>
          About the browser HAR file path requested by ArkoseLabs
  -K, --arkose-har-upload-key <ARKOSE_HAR_UPLOAD_KEY>
          HAR file upload authenticate key
  -s, --arkose-solver <ARKOSE_SOLVER>
          About ArkoseLabs solver platform [default: yescaptcha]
  -k, --arkose-solver-key <ARKOSE_SOLVER_KEY>
          About the solver client key by ArkoseLabs
  -S, --sign-secret-key <SIGN_SECRET_KEY>
          Enable url signature (signature secret key)
  -T, --tb-enable
          Enable token bucket flow limitation
      --tb-store-strategy <TB_STORE_STRATEGY>
          Token bucket store strategy (mem/redis) [default: mem]
      --tb-redis-url <TB_REDIS_URL>
          Token bucket redis connection url [default: redis://127.0.0.1:6379]
      --tb-capacity <TB_CAPACITY>
          Token bucket capacity [default: 60]
      --tb-fill-rate <TB_FILL_RATE>
          Token bucket fill rate [default: 1]
      --tb-expired <TB_EXPIRED>
          Token bucket expired (seconds) [default: 86400]
      --cf-site-key <CF_SITE_KEY>
          Cloudflare turnstile captcha site key [env: CF_SECRET_KEY=]
      --cf-secret-key <CF_SECRET_KEY>
          Cloudflare turnstile captcha secret key [env: CF_SITE_KEY=]
  -D, --disable-webui
          Disable WebUI [env: DISABLE_WEBUI=]
  -h, --help
          Print help
```

### 平台支持

- Linux
  - `x86_64-unknown-linux-musl`
  - `aarch64-unknown-linux-musl`
  - `armv7-unknown-linux-musleabi`
  - `armv7-unknown-linux-musleabihf`
  - `arm-unknown-linux-musleabi`
  - `arm-unknown-linux-musleabihf`
  - `armv5te-unknown-linux-musleabi`
- Windows
  - `x86_64-pc-windows-msvc`
- MacOS
  - `x86_64-apple-darwin`
  - `aarch64-apple-darwin`

### 编译

- Linux编译，Ubuntu机器为例:

```shell
# 本机编译
git clone https://github.com/gngpp/ninja.git && cd ninja
./build.sh

# 跨平台编译，依赖于docker(如果您可以自己解决跨平台编译依赖)，默认使用docker构建linux/windows平台
./build_cross.sh 

# 默认在Macos上构建Macos平台
os=macos ./build_cross.sh

# 编译单个平台二进制，以 aarch64-unknown-linux-musl 为例:
docker run --rm -it --user=$UID:$(id -g $USER) \
  -v $(pwd):/home/rust/src \
  -v $HOME/.cargo/registry:/root/.cargo/registry \
  -v $HOME/.cargo/git:/root/.cargo/git \
  ghcr.io/gngpp/ninja-builder:x86_64-unknown-linux-musl \
  cargo build --release
```

- OpenWrt 编译

```shell
cd package
svn co https://github.com/gngpp/ninja/trunk/openwrt
cd -
make menuconfig # choose LUCI->Applications->luci-app-ninja  
make V=s
```

### 预览

![img0](./doc/img/img0.png)
![img1](./doc/img/img1.png)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [v0.2.1](https://github.com/gngpp/ninja/releases/tag/v0.2.1) - 2023-07-21

<small>[Compare with v0.2.0](https://github.com/gngpp/ninja/compare/v0.2.0...v0.2.1)</small>

## [v0.2.0](https://github.com/gngpp/ninja/releases/tag/v0.2.0) - 2023-07-18

<small>[Compare with v0.1.9](https://github.com/gngpp/ninja/compare/v0.1.9...v0.2.0)</small>

## [v0.1.9](https://github.com/gngpp/ninja/releases/tag/v0.1.9) - 2023-07-18

<small>[Compare with v0.1.8](https://github.com/gngpp/ninja/compare/v0.1.8...v0.1.9)</small>

## [v0.1.8](https://github.com/gngpp/ninja/releases/tag/v0.1.8) - 2023-07-17

<small>[Compare with v0.1.7](https://github.com/gngpp/ninja/compare/v0.1.7...v0.1.8)</small>

### Added

- Add CHANGELOG.md ([3c25f21](https://github.com/gngpp/ninja/commit/3c25f21ce9946746e0d2d941a408f58cd5e148fe) by gngpp).

### Fixed

- fix windows build ([9ef8bce](https://github.com/gngpp/ninja/commit/9ef8bce747c932abc191ebf30849dc299c54afb2) by gngpp).

### Changed

- change template Help & FQA link ([a4c8b42](https://github.com/gngpp/ninja/commit/a4c8b429f15f4bf7c55a8306bd2bd9057daceea8) by gngpp).

## [v0.1.7](https://github.com/gngpp/ninja/releases/tag/v0.1.7) - 2023-07-15

<small>[Compare with v0.1.6](https://github.com/gngpp/ninja/compare/v0.1.6...v0.1.7)</small>

## [v0.1.6](https://github.com/gngpp/ninja/releases/tag/v0.1.6) - 2023-07-15

<small>[Compare with v0.1.5](https://github.com/gngpp/ninja/compare/v0.1.5...v0.1.6)</small>

### Added

- Adds support for ALL_PROXY ([c9b1baa](https://github.com/gngpp/ninja/commit/c9b1baa093bc3040e6d3a960d9f04c67ffa50f88) by gngpp).

## [v0.1.5](https://github.com/gngpp/ninja/releases/tag/v0.1.5) - 2023-07-12

<small>[Compare with v0.1.4](https://github.com/gngpp/ninja/compare/v0.1.4...v0.1.5)</small>

### Added

- Add x86_64-pc-windows-msvc target support ([353cc43](https://github.com/gngpp/ninja/commit/353cc435f506e4ebdc2786d7f969ab8ae089bc94) by gngpp).

## [v0.1.4](https://github.com/gngpp/ninja/releases/tag/v0.1.4) - 2023-07-11

<small>[Compare with v0.1.3](https://github.com/gngpp/ninja/compare/v0.1.3...v0.1.4)</small>

## [v0.1.3](https://github.com/gngpp/ninja/releases/tag/v0.1.3) - 2023-07-10

<small>[Compare with v0.1.2](https://github.com/gngpp/ninja/compare/v0.1.2...v0.1.3)</small>

## [v0.1.2](https://github.com/gngpp/ninja/releases/tag/v0.1.2) - 2023-07-10

<small>[Compare with v0.1.1](https://github.com/gngpp/ninja/compare/v0.1.1...v0.1.2)</small>

### Added

- Add openwrt action ([edde18a](https://github.com/gngpp/ninja/commit/edde18ae65f1961228f3bbc3bf438274e47d29a2) by gngpp).
- Add openwrt support ([71fdd5a](https://github.com/gngpp/ninja/commit/71fdd5a6c79beaf0302959d5bbe1218b49afe988) by gngpp).

## [v0.1.1](https://github.com/gngpp/ninja/releases/tag/v0.1.1) - 2023-07-08

<small>[Compare with v0.1.0](https://github.com/gngpp/ninja/compare/v0.1.0...v0.1.1)</small>

## [v0.1.0](https://github.com/gngpp/ninja/releases/tag/v0.1.0) - 2023-07-08

<small>[Compare with first commit](https://github.com/gngpp/ninja/compare/fb2d098c05b60987ca7874c72a094ad9f5ab474a...v0.1.0)</small>

### Added

- Add cargo registry ([9bc5886](https://github.com/gngpp/ninja/commit/9bc588637e1b849ff27b50335a826310441247fb) by gngpp).
- Add docker builder image ([1ace0f5](https://github.com/gngpp/ninja/commit/1ace0f5c78e741145752097cac011bb65206aa33) by gngpp).
- Add request eventsource mod ([bd5bbe0](https://github.com/gngpp/ninja/commit/bd5bbe0a426f05337323fd5d6be5d109edba2c6a) by gngpp).
- Add platform api client ([3f077ed](https://github.com/gngpp/ninja/commit/3f077ed6923d8bab9d08d5a8ce4b6141bb53632b) by gngpp).
- Add http rest doc ([43994c6](https://github.com/gngpp/ninja/commit/43994c61eb748e0808660f23d95ecd5a3c6cf5bd) by gngpp).
- Add billing credit_grants api ([fcfd054](https://github.com/gngpp/ninja/commit/fcfd054387f8da9653463a31fa9edc1eab938a71) by gngpp).
- Add billing_usage api ([b12f31f](https://github.com/gngpp/ninja/commit/b12f31fd6f293f0077d42e9c5c23091e94e151ca) by gngpp).
- Add new api ([f613183](https://github.com/gngpp/ninja/commit/f6131830d14c782826ced5ecd6f0a523a0e801d1) by gngpp).
- Add ui mod ([971d259](https://github.com/gngpp/ninja/commit/971d2594b04869d27d83851632bf982fc6cde215) by gngpp).

### Fixed

- fix decode escape convert ([51e5bd6](https://github.com/gngpp/ninja/commit/51e5bd6b09e7157cbf94d62a72acde710343719f) by gngpp).
- fix ios device api aceess ([f3b1e33](https://github.com/gngpp/ninja/commit/f3b1e33e7fe445401edf6e586199c336bd089e3d) by gngpp).

### Removed

- remove commit ([4fff81a](https://github.com/gngpp/ninja/commit/4fff81a78aa8943c7e30c5f90fe3e5ffea872ab9) by gngpp).



src/args.rs
use crate::parse;
use clap::{Args, Parser, Subcommand};
use openai::{arkose::funcaptcha::Solver, serve::tokenbucket::Strategy};
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

#[derive(Parser)]
#[clap(author, version, about, arg_required_else_help = true)]
pub(super) struct Opt {
    #[clap(subcommand)]
    pub(super) command: Option<SubCommands>,
    /// Log level (info/debug/warn/trace/error)
    #[clap(short = 'L', long, global = true, env = "LOG", default_value = "info")]
    pub(super) level: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Subcommand)]
pub(super) enum SubCommands {
    /// Start the http server
    #[clap(subcommand)]
    Serve(ServeSubcommand),
    /// Terminal interaction
    Terminal,
}

#[derive(Subcommand)]
pub(super) enum ServeSubcommand {
    /// Run the HTTP server
    Run(ServeArgs),
    /// Stop the HTTP server daemon
    #[cfg(target_family = "unix")]
    Stop,
    /// Start the HTTP server daemon
    #[cfg(target_family = "unix")]
    Start(ServeArgs),
    /// Restart the HTTP server daemon
    #[cfg(target_family = "unix")]
    Restart(ServeArgs),
    /// Status of the Http server daemon process
    #[cfg(target_family = "unix")]
    Status,
    /// Show the Http server daemon log
    #[cfg(target_family = "unix")]
    Log,
    /// Generate config template file (toml format file)
    GT {
        /// Configuration template output to file (toml format file)
        #[clap(short, long, group = "gt")]
        out: Option<PathBuf>,
    },
}

#[derive(Args, Debug, Default, Serialize, Deserialize)]
pub(super) struct ServeArgs {
    /// Configuration file path (toml format file)
    #[clap(short = 'C', long, env = "CONFIG", value_parser = parse::parse_file_path)]
    pub(super) config: Option<PathBuf>,

    /// Server Listen host
    #[clap(short = 'H', long, env = "HOST", default_value = "0.0.0.0", value_parser = parse::parse_host)]
    pub(super) host: Option<std::net::IpAddr>,

    /// Server Listen port
    #[clap(short = 'P', long, env = "PORT", default_value = "7999", value_parser = parse::parse_port_in_range)]
    pub(super) port: Option<u16>,

    /// Server worker-pool size (Recommended number of CPU cores)
    #[clap(short = 'W', long, default_value = "1")]
    pub(super) workers: usize,

    /// Enforces a limit on the concurrent number of requests the underlying
    #[clap(long, default_value = "65535")]
    pub(super) concurrent_limit: usize,

    /// Server proxies pool, Example: protocol://user:pass@ip:port
    #[clap(short = 'x',long, env = "PROXIES", value_parser = parse::parse_proxies_url)]
    pub(super) proxies: Option<std::vec::Vec<String>>,

    /// Disable direct connection
    #[clap(long, env = "DISABLE_DIRECT")]
    pub(super) disable_direct: bool,

    /// Client timeout (seconds)
    #[clap(long, default_value = "600")]
    pub(super) timeout: usize,

    /// Client connect timeout (seconds)
    #[clap(long, default_value = "60")]
    pub(super) connect_timeout: usize,

    /// TCP keepalive (seconds)
    #[clap(long, default_value = "60")]
    pub(super) tcp_keepalive: usize,

    /// TLS certificate file path
    #[clap(long, env = "TLS_CERT", requires = "tls_key")]
    pub(super) tls_cert: Option<PathBuf>,

    /// TLS private key file path (EC/PKCS8/RSA)
    #[clap(long, env = "TLS_KEY", requires = "tls_cert")]
    pub(super) tls_key: Option<PathBuf>,

    /// PUID cookie value of Plus account
    #[clap(long, env = "PUID")]
    pub(super) puid: Option<String>,

    /// Obtain the PUID of the Plus account user, Example: `user:pass` or `user:pass:mfa`
    #[clap(long, value_parser = parse::parse_puid_user)]
    pub(super) puid_user: Option<(String, String, Option<String>)>,

    /// Web UI api prefix
    #[clap(long, env = "UI_API_PREFIX", value_parser = parse::parse_url)]
    pub(super) api_prefix: Option<String>,

    /// Arkose endpoint, Example: https://client-api.arkoselabs.com
    #[clap(long, value_parser = parse::parse_url)]
    pub(super) arkose_endpoint: Option<String>,

    /// Get arkose token endpoint
    #[clap(short = 'A', long, value_parser = parse::parse_url)]
    pub(super) arkose_token_endpoint: Option<String>,

    /// About the browser HAR file path requested by ArkoseLabs
    #[clap(short, long, value_parser = parse::parse_file_path)]
    pub(super) arkose_har_file: Option<PathBuf>,

    /// HAR file upload authenticate key
    #[clap(short = 'K', long)]
    pub(super) arkose_har_upload_key: Option<String>,

    /// About ArkoseLabs solver platform
    #[clap(
        short = 's',
        long,
        default_value = "yescaptcha",
        requires = "arkose_solver_key"
    )]
    pub(super) arkose_solver: Solver,

    #[clap(short = 'k', long)]
    /// About the solver client key by ArkoseLabs
    pub(super) arkose_solver_key: Option<String>,

    /// Enable url signature (signature secret key)
    #[clap(short = 'S', long)]
    #[cfg(feature = "sign")]
    pub(super) sign_secret_key: Option<String>,

    /// Enable token bucket flow limitation
    #[clap(short = 'T', long)]
    #[cfg(feature = "limit")]
    pub(super) tb_enable: bool,

    /// Token bucket store strategy (mem/redis)
    #[clap(long, default_value = "mem", requires = "tb_enable")]
    #[cfg(feature = "limit")]
    pub(super) tb_store_strategy: Strategy,

    /// Token bucket redis connection url
    #[clap(long, default_value = "redis://127.0.0.1:6379", requires = "tb_enable", value_parser = parse::parse_url)]
    #[cfg(feature = "limit")]
    pub(super) tb_redis_url: String,

    /// Token bucket capacity
    #[clap(long, default_value = "60", requires = "tb_enable")]
    #[cfg(feature = "limit")]
    pub(super) tb_capacity: u32,

    /// Token bucket fill rate
    #[clap(long, default_value = "1", requires = "tb_enable")]
    #[cfg(feature = "limit")]
    pub(super) tb_fill_rate: u32,

    /// Token bucket expired (seconds)
    #[clap(long, default_value = "86400", requires = "tb_enable")]
    #[cfg(feature = "limit")]
    pub(super) tb_expired: u32,

    /// Cloudflare turnstile captcha site key
    #[clap(long, env = "CF_SECRET_KEY", requires = "cf_secret_key")]
    pub(super) cf_site_key: Option<String>,

    /// Cloudflare turnstile captcha secret key
    #[clap(long, env = "CF_SITE_KEY", requires = "cf_site_key")]
    pub(super) cf_secret_key: Option<String>,

    /// Disable WebUI
    #[clap(short = 'D', long, env = "DISABLE_WEBUI")]
    pub(super) disable_webui: bool,
}
